onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,371 @@
+void QlinearBuildLookupTableUint32(uint32_t* table,
+                                   const float x_scale,
+                                   size_t reduce_len) {
+  static_assert(std::is_same<T, int8_t>::value ||
+                    std::is_same<T, uint8_t>::value,
+                "unsupported quant type");
+  const double qscale =
+      fmin(static_cast<double>(UINT32_MAX) / static_cast<double>(reduce_len), static_cast<double>(0x7fffff));
+  for (int32_t i = 0; i < 256; i++) {
+    double scaled_exp_xi = qscale * exp(static_cast<double>(i - 255) * static_cast<double>(x_scale));
+    // we can't get the real max number of input tensor here, so we just assume 255.
+    // in the process of computation, all numbers will have a shift to align 255
+    if constexpr (std::is_same<T, int8_t>::value) {
+      // 1 2 3 ......126 127 -128 -127 ..... -3 -2 -1
+      uint8_t index = static_cast<uint8_t>(i - 128);
+      table[index] = static_cast<uint32_t>(lrint(scaled_exp_xi));
+    } else {
+      table[i] = static_cast<uint32_t>(lrint(scaled_exp_xi));
+    }
+  }
+}
+}  // namespace
+
+template <typename T>
+void QLinearSoftmax<T>::BuildLookupTableIfFixed(const OpKernelInfo& info, uint32_t reduce_len) {
+  const Tensor* tensor_x_scale = nullptr;
+
+  bool get_x_scale = info.TryGetConstantInput(1, &tensor_x_scale);
+  ORT_ENFORCE(tensor_x_scale == nullptr || IsScalarOr1ElementVector(tensor_x_scale),
+              "QlinearBuildLookupTable : input X_scale must be a scalar or 1D tensor of size 1");
+  bool is_fixed_parameters = get_x_scale;
+
+  if (is_fixed_parameters) {
+    fixed_lookup_table_.resize(256);
+    const float X_scale = *(tensor_x_scale->Data<float>());
+    QlinearBuildLookupTableUint32<T>(fixed_lookup_table_.data(), X_scale, reduce_len);
+  }
+}
+
+template <typename T>
+QLinearSoftmax<T>::QLinearSoftmax(const OpKernelInfo& info)
+    : OpKernel(info) {
+  const auto& node = info.node();
+  int64_t opset = -1;
+  Status status = info.GetAttr<int64_t>("opset", &opset);
+  ORT_ENFORCE(status.IsOK(), "opset must be existed in attributes of QlinearSoftmax");
+  opset_ = gsl::narrow_cast<int>(opset);
+
+  int64_t axis = -1;
+  status = info.GetAttr<int64_t>("axis", &axis);
+  if (status.IsOK()) {
+    axis_ = gsl::narrow_cast<int>(axis);
+  } else {
+    if (opset_ < OPSET13) {
+      axis_ = 1;  // opset-12 and below, the default axis value is 1
+    } else {
+      axis_ = -1;  // opset-13, the default axis value is -1
+    }
+  }
+  auto input_defs = node.InputDefs();
+  const auto& x_shape = input_defs[0]->Shape();
+  int rank = x_shape->dim_size();
+  if (rank == 0) {
+    return;
+  }
+  if (axis_ < 0) {
+    axis_ = static_cast<int>(HandleNegativeAxis(axis_, int64_t(rank)));
+  }
+  uint32_t reduce_size = gsl::narrow_cast<uint32_t>(x_shape->dim(axis_).dim_value());
+  if (opset_ < OPSET13) {
+    for (int i = axis_ + 1; i < rank; i++) {
+      reduce_size *= gsl::narrow_cast<uint32_t>(x_shape->dim(i).dim_value());
```

Comment:
TensorShape has a function to do this.

Reply:
`X_shape` is  a var of `TensorShapeProto`, seems we don't have a help function to transform from `TensorShapeProto` to `TensorShape`.

Reply:
GetTensorShapeFromTensorShapeProto

Reply:
Thanks